### PR TITLE
🏗🐛 Prevent most async throwing of errors during tests

### LINKED ISSUE #15658: Mocha silently skips tests when an async error is thrown
@jpettitt reported this while running the following test:

https://github.com/ampproject/amphtml/blob/97847f23b36b5004a63bdc94f35700bfb48c5ac4/extensions/amp-user-notification/0.1/test/test-amp-user-notification.js#L80-L88

The error at the end of the test gets thrown asynchronously, by which time the test has completed. So Mocha makes it look like the test has passed, but the thrown error prevents it from running subsequent tests.

See https://github.com/mochajs/mocha/issues/3223 and https://github.com/mochajs/mocha/pull/3257 for more info on why this is a mocha problem.

### PR DESCRIPTION
**Pull request by rsimha**
This PR does the following:
1. Stubs out `rethrowAsync` in `src/log.js` during tests with `rethrowAsyncForTests`
2. Creates a mechanism with which tests can opt in to the default async throwing behavior (for example, `test-log.js`)

With this. the vast majority of async throw problems we're seeing should be eliminated.

Partial fix for #14406
Partial fix for #14360
Fixes #15658

### CODE REVIEW

**Comment by rsimha** [comment]
No, it's 19 failures out of 400 tests. Most of the 8000 tests didn't run. This isn't ready yet.

File: src/log.js [hunk 1]
```diff
@@ -548,6 +548,19 @@ export function rethrowAsync(var_args) {
 }
 
 
+/**
+ * Rethrows the error immediately. Required because async errors thrown during
+ * tests leak to subsequent tests. DO NOT USE IN RUNTIME CODE.
+ * @visibleForTesting
+ * @param {...*} var_args
+ */
+export function rethrowAsyncForTests(var_args) {
```
Comment (erwinmombay) [inline src/log.js]: we allow rest/spread in the code base right? lets just use that instead of var_args

Comment (rsimha) [reply src/log.js]: We use `var_args` in several other spots in `log.js`, so I wanted to be consistent here. Also, there's a lint rule against using the spread operator.

https://github.com/ampproject/amphtml/blob/db751cb6e9db35cda235484cf6e8f5708a41566b/.eslintrc#L74

If it's possible to use rest/spread here, I think it makes sense to switch this and all other instances in the file in a separate PR.

Comment (lannka) [inline src/log.js]: can we move this method to init_tests? or any where under test folders.

File: test/functional/test-3p.js [hunk 2]
```diff
@@ -147,14 +147,12 @@ describe('3p', () => {
...
 
-      validateData({
-        type: 'TEST',
-        foo: true,
-        'not-whitelisted': true,
-      }, [], ['foo']);
-
       allowConsoleError(() => { expect(() => {
-        clock.tick(1);
+        validateData({
```
Comment (lannka) [inline test/functional/test-3p.js]: i guess you can remove all the clock ticks above in this test.

File: src/log.js (hunk 1 above)
Comment (rsimha) [reply src/log.js]: I meant to do this once I got everything to work. Done.

File: test/functional/test-3p.js (hunk 2 above)
Comment (rsimha) [reply test/functional/test-3p.js]: Done.

File: test/functional/test-xhr.js [hunk 3]
```diff
@@ -381,7 +381,8 @@ describe.configure().skipSafari().run('XHR', function() {
         });
       });
 
-      it('should do simple JSON fetch', () => {
+      // TODO(#16916): Make this test work with synchronous throws.
+      it.skip('should do simple JSON fetch', () => {
```
Comment (prateekbh) [inline test/functional/test-xhr.js]: I have no explanation why but after this, it silently fails rest of the tests in the file.

What should have been an execution of 100 tests altogether, are now just 9 tests but passing

results via locally running `gulp test --files=test/functional/test-xhr.js --watch`

Comment (rsimha) [reply test/functional/test-xhr.js]: If you hit the debug button in the karma window, do you see any errors when the tests are re run?

Comment (rsimha) [reply test/functional/test-xhr.js]: @prateekbh I was able to repro your problem. One of your tests is throwing an error in `beforeEach()`. Mocha considers this a catastrophic error and bails. Let me know if you need help reproing it.

Comment (prateekbh) [reply test/functional/test-xhr.js]: well i can use some help on how to find whats the issue, I am creating a PR which just toucher `xhr-impl.js` and hence I reproduced this error. 
Not really sure about the setup and test cases here in `functional`

**Pull request merged**